src/cheri_insts.sail

```diff
@@ -745,70 +769,86 @@ function CSeal_core (cd, cs1, cs1_val, cs2, cs2_val) = {
-}
-
 union clause ast = CCSeal : (regidx, regidx, regidx)
+/*!
+ * Capability register *cd* is replaced with capability register *cs1*, and is
+ * conditionally sealed with **otype** equal to the **address** field of
+ * capability register *cs2*. The conditions under which the input is passed
+ * through unaltered are intended to permit a fast branchless rederivation
+ * sequence with multiple sealing authorities with a single [CBuildCap] and a
+ * set of [CCopyType] and [CCSeal] pairs when swapping capabilities in from
+ * disk.
+ *
+ * ## Exceptions
+ *
+ * An exception is raised if:
+ *   - *cs1*.**tag** is not set.
+ *   - *cs2*.**sealed** is set.
+ *   - *cs2*.**perms** does not grant Permit_Seal.
+ *   - *cs2*.**address** $\gt$ [max_otype].
+ *
+ * ## Notes
+ *
+ * - Since the intent is that this is used for rederiving swapped-out
+ *   capabilities, the expectation is that this whole sequence is guarded by a
+ *   check on whether the **tag** field of the capability was valid, and so any
+ *   invalid capability input in *cs1* will cause a trap.
+ *
+ * - If the input to be conditionally sealed is already sealed it is passed
+ *   through before any futher checks are made. This allows multiple [CCSeal]s
+ *   in a chain, any of which can be the one to seal the initial input. The
+ *   intent is that all of these [CCSeal]s' authorities will have been produced
+ *   by [CCopyType]s of the same input (i.e., they will all attempt to seal to
+ *   the same type), but that's not, strictly, required. Sealed capabilities
+ *   with a reserved **otype** are also constructed directly by [CBuildCap].
+ *
+ * - To avoid the need to branch on whether the original capability was sealed,
+ *   attempts to seal with the reserved unsealed **otype** will leave the
+ *   capability unmodified rather than trap.
+ *
+ * - To avoid the need to check which is the correct authority, any sealing
+ *   request where the **address** of capability register *cs2* is out of
+ *   bounds will leave the capability unmodified rather than trap, as will
+ *   attempts to seal with an invalid capability since it may have become
+ *   unrepresentable but be within its reinterpreted bounds.
+ */
 function clause execute (CCSeal(cd, cs1, cs2)) = {
-  let cs1_val = C(cs1); /* Tagged but unsealed cap */
-  let cs2_val = C(cs2); /* Authority */
+  let cs1_val = C(cs1);
+  let cs2_val = C(cs2);
   let cs2_cursor = getCapCursor(cs2_val);
-
-    /*
-     * The "conditional" aspects of CCSeal mean that we do some tests on the
-     * thing to be sealed before we actually check the authority.
-     */
+  let (cs2_base, cs2_top) = getCapBounds(cs2_val);
   if not (cs1_val.tag) then {
     handle_cheri_reg_exception(CapEx_TagViolation, cs1);
     RETIRE_FAIL
+  } else if not (cs2_val.tag) then {
+    /* CCopyType may not have been able to represent the result */
+    C(cd) = cs1_val;
+    RETIRE_SUCCESS
   } else if cs1_val.sealed then {
```

Comment:
Unrelated to this PR but maybe we should remove the separate sealed field and use a helper function instead? It made sense for 256 and 128 prior to the new format but I think now it's more likely to cause bugs then be useful.

Reply:
Yeah it's a bug waiting to happen (and ISTR Peter did fall afoul of that recently?).

Reply:
Indeed, prompted the ugly "cap with otype=unsealed, sealed=false"

Reply:
Done in f27e6397da